File: project2context.py
import os
import sys
from pathlib import Path

# Configuration
EXCLUDED_DIRS = {
    'node_modules',
    'venv',
    '.git',
    '__pycache__',
    'build',
    'dist'
}

EXCLUDED_FILES = {
    'package-lock.json',
    'yarn.lock',
    'project-context.txt'
}

TEXT_EXTENSIONS = {
    '.txt', '.md', '.py', '.js', '.ts', '.jsx', '.tsx', '.cpp', '.h', 
    '.hpp', '.c', '.cs', '.java', '.html', '.css', '.scss', '.sass',
    '.json', '.yml', '.yaml', '.xml', '.env', '.config', '.dockerfile',
    '.sh', '.bat', '.ps1'
}

EXCLUDED_EXTENSIONS = {
    '.svg', '.png', '.jpg', '.jpeg', '.gif', '.ico', '.pdf', '.zip',
    '.tar', '.gz', '.rar', '.exe', '.dll', '.pdb', '.pyc'
}

def get_script_name():
    """Returns the name of the currently executing script."""
    return os.path.basename(sys.argv[0]) if sys.argv[0].endswith('.py') else os.path.basename(__file__)

def is_text_file(file_path):
    """Checks if a file is a text file by looking for null bytes."""
    try:
        with open(file_path, 'rb') as file:
            chunk = file.read(1024)
            return b'\0' not in chunk
    except Exception:
        return False

def get_file_encoding(file_path):
    """Detects file encoding by checking for BOM markers."""
    with open(file_path, 'rb') as file:
        raw = file.read(4)
        if raw.startswith(b'\xef\xbb\xbf'):
            return 'utf-8-sig'
        if raw.startswith(b'\xff\xfe') or raw.startswith(b'\xfe\xff'):
            return 'utf-16'
        if raw.startswith(b'\xff\xfe\x00\x00') or raw.startswith(b'\x00\x00\xfe\xff'):
            return 'utf-32'
        return 'utf-8'

def estimate_tokens(text):
    """Rough estimation of token count based on word count + special characters."""
    words = len(text.split())
    special_chars = len([c for c in text if not c.isalnum() and not c.isspace()])
    return words + special_chars + (words // 4)

def should_process_file(file_path):
    """Determines if a file should be processed based on configuration."""
    path = Path(file_path)
    
    if path.name == get_script_name():
        return False
        
    if path.name in EXCLUDED_FILES:
        return False
    if path.suffix.lower() in EXCLUDED_EXTENSIONS:
        return False
    if not (path.suffix.lower() in TEXT_EXTENSIONS or path.suffix == ''):
        return False
    
    return is_text_file(file_path)

def generate_tree(start_path, output_file, prefix=""):
    """Generates a directory tree structure."""
    if os.path.basename(start_path) in EXCLUDED_DIRS:
        return

    output_file.write(f"{prefix}{os.path.basename(start_path)}/\n")
    
    try:
        items = sorted(os.scandir(start_path), key=lambda x: (not x.is_dir(), x.name.lower()))
    except PermissionError:
        output_file.write(f"{prefix}[ACCESS DENIED]\n")
        return
    
    for i, item in enumerate(items):
        is_last = i == len(items) - 1
        current_prefix = prefix + ("└─ " if is_last else "├─ ")
        next_prefix = prefix + ("   " if is_last else "│  ")
        
        if item.is_dir():
            if item.name not in EXCLUDED_DIRS:
                generate_tree(item.path, output_file, next_prefix)
        else:
            if should_process_file(item.path):
                output_file.write(f"{current_prefix}{item.name}\n")

def process_directory(start_path, output_path, base_dir):
    """Main function for processing the project directory."""
    total_lines = 0
    total_tokens = 0
    file_count = 0
    
    with open(output_path, 'w', encoding='utf-8') as output_file:
        output_file.write("<project_overview>\n")
        output_file.write(f"<generated_at>{os.path.basename(start_path)}</generated_at>\n\n")
        
        output_file.write("<directory_structure>\n")
        generate_tree(start_path, output_file)
        output_file.write("</directory_structure>\n\n")
        
        output_file.write("<file_contents>\n")
        
        for root, dirs, files in os.walk(start_path):
            dirs[:] = [d for d in dirs if d not in EXCLUDED_DIRS]
            
            for file in sorted(files, key=str.lower):
                file_path = os.path.join(root, file)
                
                if should_process_file(file_path):
                    try:
                        encoding = get_file_encoding(file_path)
                        with open(file_path, 'r', encoding=encoding) as f:
                            content = f.read()
                            
                        # Use relative path from the base directory
                        rel_path = os.path.relpath(file_path, base_dir)
                        output_file.write(f"\n<file path=\"{rel_path}\">\n")
                        output_file.write(content)
                        output_file.write("\n</file>\n")
                        
                        total_lines += content.count('\n') + 1
                        total_tokens += estimate_tokens(content)
                        file_count += 1
                        
                    except Exception as e:
                        # Use relative path from the base directory
                        rel_path = os.path.relpath(file_path, base_dir)
                        output_file.write(f"\n<error file=\"{rel_path}\">{str(e)}</error>\n")

        output_file.write("</file_contents>\n")
        output_file.write("</project_overview>")

    return file_count, total_lines, total_tokens

def parse_args():
    """Parse command line arguments."""
    args = sys.argv[1:]
    
    # Default value
    options = {
        'target_dir': ''
    }
    
    # Simple parsing: treat the first argument as the target directory
    if len(args) > 0:
        options['target_dir'] = args[0]
    
    return options

if __name__ == "__main__":
    try:
        # Parse command line args
        options = parse_args()
        
        # Fixed output file name as per README
        output_file = "project-context.txt"
        
        # Get the current directory
        current_dir = os.getcwd()
        
        # Resolve the target directory
        target_dir = options['target_dir']
        resolved_target_dir = os.path.join(current_dir, target_dir) if target_dir else current_dir
        
        # Make sure the directory exists
        if not os.path.exists(resolved_target_dir):
            raise FileNotFoundError(f"Directory does not exist: {resolved_target_dir}")
        
        if not os.path.isdir(resolved_target_dir):
            raise NotADirectoryError(f"Not a directory: {resolved_target_dir}")
        
        # Base directory for relative paths in output (current dir)
        base_dir = current_dir
        
        print(f"Processing directory: {os.path.relpath(resolved_target_dir, current_dir) or '.'}")
        
        files, lines, tokens = process_directory(resolved_target_dir, output_file, base_dir)
        
        print(f"\nProject Context Generation Complete")
        print(f"Output file: {output_file}")
        print(f"Files processed: {files}")
        print(f"Total lines: {lines}")
        print(f"Estimated tokens: {tokens}")
        
    except Exception as e:
        print(f"Error: {str(e)}")

File: project2context.js
const fs = require('fs');
const path = require('path');

// Configuration
const EXCLUDED_DIRS = new Set([
    'node_modules',
    'venv',
    '.git',
    '__pycache__',
    'build',
    'dist'
]);

const EXCLUDED_FILES = new Set([
    'package-lock.json',
    'yarn.lock',
    'project-context.txt'
]);

const TEXT_EXTENSIONS = new Set([
    '.txt', '.md', '.py', '.js', '.ts', '.jsx', '.tsx', '.cpp', '.h',
    '.hpp', '.c', '.cs', '.java', '.html', '.css', '.scss', '.sass',
    '.json', '.yml', '.yaml', '.xml', '.env', '.config', '.dockerfile',
    '.sh', '.bat', '.ps1'
]);

const EXCLUDED_EXTENSIONS = new Set([
    '.svg', '.png', '.jpg', '.jpeg', '.gif', '.ico', '.pdf', '.zip',
    '.tar', '.gz', '.rar', '.exe', '.dll', '.pdb', '.pyc'
]);

function getScriptName()
{
    /**
     * Returns the name of the currently executing script.
     */
    const scriptPath = process.argv[1];
    return path.basename(scriptPath);
}

function isTextFile(filePath)
{
    /**
     * Checks if a file is a text file by looking for null bytes.
     */
    try
    {
        const buffer = fs.readFileSync(filePath, { encoding: 'binary', flag: 'r' }).slice(0, 1024);
        return !buffer.includes('\0');
    } catch (error)
    {
        return false;
    }
}

function getFileEncoding(filePath)
{
    /**
     * Detects file encoding by checking for BOM markers.
     */
    try
    {
        const buffer = Buffer.alloc(4);
        const fd = fs.openSync(filePath, 'r');
        fs.readSync(fd, buffer, 0, 4, 0);
        fs.closeSync(fd);

        if (buffer[0] === 0xEF && buffer[1] === 0xBB && buffer[2] === 0xBF)
        {
            return 'utf8';  // UTF-8 with BOM
        }
        if (buffer[0] === 0xFF && buffer[1] === 0xFE || buffer[0] === 0xFE && buffer[1] === 0xFF)
        {
            return 'utf16le';  // UTF-16
        }
        if ((buffer[0] === 0xFF && buffer[1] === 0xFE && buffer[2] === 0x00 && buffer[3] === 0x00) ||
            (buffer[0] === 0x00 && buffer[1] === 0x00 && buffer[2] === 0xFE && buffer[3] === 0xFF))
        {
            return 'utf32le';  // UTF-32
        }
        return 'utf8';  // Default to UTF-8
    } catch (error)
    {
        return 'utf8';  // Default to UTF-8 on error
    }
}

function estimateTokens(text)
{
    /**
     * Rough estimation of token count based on word count + special characters.
     */
    const words = text.split(/\s+/).length;
    const specialChars = text.split('').filter(c => !(/[a-zA-Z0-9\s]/).test(c)).length;
    return words + specialChars + Math.floor(words / 4);
}

function shouldProcessFile(filePath)
{
    /**
     * Determines if a file should be processed based on configuration.
     */
    const fileName = path.basename(filePath);
    const extension = path.extname(filePath).toLowerCase();

    if (fileName === getScriptName())
    {
        return false;
    }

    if (EXCLUDED_FILES.has(fileName))
    {
        return false;
    }

    if (EXCLUDED_EXTENSIONS.has(extension))
    {
        return false;
    }

    if (!(TEXT_EXTENSIONS.has(extension) || extension === ''))
    {
        return false;
    }

    return isTextFile(filePath);
}

function generateTree(startPath, outputFile, prefix = "")
{
    /**
     * Generates a directory tree structure.
     */
    const baseName = path.basename(startPath);

    if (EXCLUDED_DIRS.has(baseName))
    {
        return;
    }

    outputFile.write(`${prefix}${baseName}/\n`);

    let items;
    try
    {
        items = fs.readdirSync(startPath, { withFileTypes: true })
            .sort((a, b) =>
            {
                // Sort directories first, then alphabetically
                if (a.isDirectory() && !b.isDirectory()) return -1;
                if (!a.isDirectory() && b.isDirectory()) return 1;
                return a.name.toLowerCase().localeCompare(b.name.toLowerCase());
            });
    } catch (error)
    {
        outputFile.write(`${prefix}[ACCESS DENIED]\n`);
        return;
    }

    for (let i = 0; i < items.length; i++)
    {
        const item = items[i];
        const isLast = i === items.length - 1;
        const currentPrefix = prefix + (isLast ? "└─ " : "├─ ");
        const nextPrefix = prefix + (isLast ? "   " : "│  ");

        if (item.isDirectory())
        {
            if (!EXCLUDED_DIRS.has(item.name))
            {
                generateTree(path.join(startPath, item.name), outputFile, nextPrefix);
            }
        } else
        {
            const itemPath = path.join(startPath, item.name);
            if (shouldProcessFile(itemPath))
            {
                outputFile.write(`${currentPrefix}${item.name}\n`);
            }
        }
    }
}

function processDirectory(startPath, outputPath)
{
    /**
     * Main function for processing the project directory.
     */
    let totalLines = 0;
    let totalTokens = 0;
    let fileCount = 0;

    const outputFile = fs.createWriteStream(outputPath, { encoding: 'utf8' });

    outputFile.write("<project_overview>\n");
    outputFile.write(`<generated_at>${path.basename(startPath)}</generated_at>\n\n`);

    outputFile.write("<directory_structure>\n");
    generateTree(startPath, outputFile);
    outputFile.write("</directory_structure>\n\n");

    outputFile.write("<file_contents>\n");

    // Walk directory recursively
    function walkDir(dir)
    {
        const entries = fs.readdirSync(dir, { withFileTypes: true });

        // Process directories
        const dirs = entries
            .filter(entry => entry.isDirectory())
            .filter(entry => !EXCLUDED_DIRS.has(entry.name))
            .map(entry => entry.name);

        // Process files
        const files = entries
            .filter(entry => entry.isFile())
            .map(entry => entry.name)
            .sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));

        for (const file of files)
        {
            const filePath = path.join(dir, file);

            if (shouldProcessFile(filePath))
            {
                try
                {
                    const encoding = getFileEncoding(filePath);
                    const content = fs.readFileSync(filePath, { encoding });

                    const relPath = path.relative(startPath, filePath);
                    outputFile.write(`\n<file path="${relPath}">\n`);
                    outputFile.write(content);
                    outputFile.write("\n</file>\n");

                    totalLines += (content.match(/\n/g) || []).length + 1;
                    totalTokens += estimateTokens(content);
                    fileCount += 1;

                } catch (error)
                {
                    const relPath = path.relative(startPath, filePath);
                    outputFile.write(`\n<error file="${relPath}">${error.message}</error>\n`);
                }
            }
        }

        // Recursively process subdirectories
        for (const subdir of dirs)
        {
            walkDir(path.join(dir, subdir));
        }
    }

    walkDir(startPath);

    outputFile.write("</file_contents>\n");
    outputFile.write("</project_overview>");

    outputFile.end();

    return { fileCount, totalLines, totalTokens };
}

// Main execution
try
{
    const currentDir = process.cwd();
    const outputFile = "project-context.txt";
    const { fileCount, totalLines, totalTokens } = processDirectory(currentDir, outputFile);

    console.log("\nProject Context Generation Complete");
    console.log(`Output file: ${outputFile}`);
    console.log(`Files processed: ${fileCount}`);
    console.log(`Total lines: ${totalLines}`);
    console.log(`Estimated tokens: ${totalTokens}`);

} catch (error)
{
    console.error(`Error: ${error.message}`);
}